File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0, right = height.size() - 1, g = 0, area = 0, sub = 0;

        while (left < right) {
            if (min(height[right], height[left]) > g) {
                area += (right - 1 - left) * (min(height[right], height[left]) - g);
                sub = min(height[right], height[left]);
            }

            g = max(min(height[right], height[left]), sub);

            if (height[left] < height[right]) {
                left++;
                if (left < right) {
                    if (height[left] > g && left < right) {
                        area -= sub;
                    } else {
                        area -= height[left];
                    }
                }
            } else {
                right--;
                if (left < right) {
                    if (height[right] > g) {
                        area -= sub;
                    } else {
                        area -= height[right];
                    }
                }
            }
        }

        return area;
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean a = true;
    public boolean isBalanced(TreeNode root) {
        lengthOfNode(root);
        return a;
    }
    public int lengthOfNode(TreeNode root){
        if(root==null){return 0;}

        int left = lengthOfNode(root.left);
        int right = lengthOfNode(root.right);

        if(Math.abs(left-right)>1){
            a = false;
        }

        return Math.max(left,right)+1;
    }
}
